crates/oxc_linter/src/rules/eslint/dot_notation.rs

```diff
@@ -0,0 +1,343 @@
+);
+
+/// List of ES3 keywords.
+const KEYWORDS: &[&str] = &[
+    "abstract",
+    "boolean",
+    "break",
+    "byte",
+    "case",
+    "catch",
+    "char",
+    "class",
+    "const",
+    "continue",
+    "debugger",
+    "default",
+    "delete",
+    "do",
+    "double",
+    "else",
+    "enum",
+    "export",
+    "extends",
+    "false",
+    "final",
+    "finally",
+    "float",
+    "for",
+    "function",
+    "goto",
+    "if",
+    "implements",
+    "import",
+    "in",
+    "instanceof",
+    "int",
+    "interface",
+    "long",
+    "native",
+    "new",
+    "null",
+    "package",
+    "private",
+    "protected",
+    "public",
+    "return",
+    "short",
+    "static",
+    "super",
+    "switch",
+    "synchronized",
+    "this",
+    "throw",
+    "throws",
+    "transient",
+    "true",
+    "try",
+    "typeof",
+    "var",
+    "void",
+    "volatile",
+    "while",
+    "with",
+];
+lazy_static! {
+    static ref VALID_IDENTIFER: Regex = Regex::new(r"^[a-zA-Z_$][a-zA-Z0-9_$]*$").unwrap();
+}
+
+impl Rule for DotNotation {
+    fn from_configuration(value: serde_json::Value) -> Self {
+        let mut cfg = DotNotationConfig::default();
+
+        if let Some(config) = value.get(0) {
+            if let Some(val) = config.get("allowPattern").and_then(serde_json::Value::as_str) {
+                cfg.allow_pattern = RegexBuilder::new(val).build().ok();
+            }
+            if let Some(val) = config.get("allowKeywords").and_then(serde_json::Value::as_bool) {
+                cfg.allow_keywords = val;
+            }
+        }
+
+        Self(Box::new(cfg))
+    }
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        if let AstKind::MemberExpression(node) = &node.kind() {
+            match &node {
+                MemberExpression::ComputedMemberExpression(expr) => {
+                    let inner_expr =
+                        if let Expression::ParenthesizedExpression(inner_expr) = &expr.expression {
+                            &inner_expr.expression
+                        } else {
+                            &expr.expression
+                        };
+
+                    if let Some((before, after)) = match &inner_expr {
+                        Expression::NullLiteral(x) => Some((x.to_string(), x.to_string())),
+                        Expression::BooleanLiteral(x) => Some((x.to_string(), x.to_string())),
+                        Expression::StringLiteral(x) => Some((x.to_string(), x.to_string())),
+                        Expression::TemplateLiteral(x) => {
+                            if x.expressions.len() == 0 && x.quasis.len() == 1 {
+                                let s = x.quasis[0].value.cooked.unwrap().into_string();
+                                Some((s.clone(), s))
+                            } else {
+                                None
+                            }
+                        }
+                        _ => None,
+                    } {
```

Comment:
1. we have a method for getting the expression inside called `get_inner_expression`
2. there is no need to clone the strings inside, I don't think we have a helper that currently does what this is doing, but it can at least be simplified:

```suggestion
                    let value = match expr.expression.get_inner_expression() {
                    Expression::NullLiteral(x) => Some(x.to_string()),
                    Expression::BooleanLiteral(x) => Some(x.to_string()),
                    Expression::StringLiteral(x) => Some(x.to_string()),
                    Expression::TemplateLiteral(x) => {
                        if x.expressions.len() == 0 && x.quasis.len() == 1 {
                            Some(x.quasis[0].value.cooked.unwrap().to_string())
                        } else {
                            None
                        }
                    }
                    _ => None,
                };
                let Some(value) = value else {
                    return;
                };
```